rpc/grpc/server/src/connection.rs

```diff
@@ -57,46 +541,66 @@ impl Connection for GrpcConnection {
 
     fn send(&self, message: Self::Message) -> Result<(), Self::Error> {
         match !self.is_closed() {
-            true => Ok(self.inner.sender.try_send((*message).clone())?),
+            true => Ok(self.inner.outgoing_route.try_send((*message).clone())?),
             false => Err(NotificationError::ConnectionClosed.into()),
         }
     }
 
     fn close(&self) -> bool {
-        // TODO: actually close sender
-        self.inner.closed.store(true, Ordering::SeqCst);
-        true
+        if let Some(signal) = self.inner.shutdown_signal.lock().take() {
+            let _ = signal.send(());
+            true
+        } else {
+            // This means the connection was already closed
+            debug!("gRPC: Connection close was called more than once, client-id: {}", self.identity());
+            false
+        }
     }
 
     fn is_closed(&self) -> bool {
-        self.inner.sender.is_closed() || self.inner.closed.load(Ordering::SeqCst)
+        self.inner.shutdown_signal.lock().is_none()
     }
 }
 
-pub(crate) struct GrpcConnectionManager {
-    connections: HashMap<SocketAddr, GrpcConnection>,
-    notifier: Arc<Notifier<Notification, GrpcConnection>>,
+#[derive(Clone)]
+pub struct GrpcConnectionManager {
+    connections: Arc<Mutex<HashMap<SocketAddr, GrpcConnection>>>,
+    max_connections: usize,
 }
 
 impl GrpcConnectionManager {
-    pub fn new(notifier: Arc<Notifier<Notification, GrpcConnection>>) -> Self {
-        Self { connections: HashMap::new(), notifier }
+    pub fn new(max_connections: usize) -> Self {
+        Self { connections: Arc::new(Mutex::new(HashMap::new())), max_connections }
     }
 
-    pub fn register(&mut self, address: SocketAddr, sender: GrpcSender) -> ListenerId {
-        let connection = GrpcConnection::new(address, sender);
-        let id = self.notifier.clone().register_new_listener(connection.clone());
-        trace!("registering a new gRPC connection from: {address} with listener id {id}");
+    pub fn register(&self, connection: GrpcConnection) {
+        debug!("gRPC: Register a new connection from {connection}");
+        let prev_connection = self.connections.lock().insert(connection.identity(), connection);
+        if let Some(prev) = prev_connection {
+            if !prev.is_closed() {
+                prev.close();
```

Comment:
why not just call `close()` w/o checking, since duplicate calls are handled within